### PR TITLE
Update Prometheus & TSDB

### PR DESCRIPTION
**Pull request by tomwilkie**
Update vendored Prometheus; lots of changes, too many to enumerate.

Adaptations in Cortex to deal with it:
- Querier.Select now returns a third type, warnings.  Just return nil in all cases.
- Querier.LabelNames is needed; return nil for now, until we implement it.
- promql.ErrStorage has changed.  Update for that.
- promql.NewEngine takes a struct; also new option for limiting # samples in a query.

The updated Prometheus dropped the parsing of the old alert and rule syntax; we copy the old promql engine code into Cortex to continue to be able to parse them.

### CODE REVIEW

**Comment by tomwilkie** [comment]
See https://github.com/prometheus/prometheus/issues/4956 for some of the lint failures...

**Comment by bboreham** [comment]
The change in `ErrStorage` should improve some of the "error is 500 when we expected 400" issues, I think.
Maybe #780 ?

File: pkg/ruler/ruler.go [hunk 1]
```diff
@@ -249,7 +249,7 @@ func buildNotifierConfig(rulerConfig *Config) (*config.Config, error) {
 	amConfig := &config.AlertmanagerConfig{
 		Scheme:                 u.Scheme,
 		PathPrefix:             u.Path,
-		Timeout:                rulerConfig.NotificationTimeout,
+		Timeout:                model.Duration(rulerConfig.NotificationTimeout),
```
[the remark this answers is not in the document]
Comment (tomwilkie) [reply pkg/ruler/ruler.go]: Yes, moved to the "Keep up with changes in the Prometheus API." cset.

File: pkg/ruler/ruler_test.go [hunk 2]
```diff
@@ -28,7 +28,11 @@ func newTestRuler(t *testing.T, alertmanagerURL string) *Ruler {
 	// TODO: Populate distributor and chunk store arguments to enable
 	// other kinds of tests.
 
-	engine := promql.NewEngine(nil, nil, 20, 2*time.Minute)
+	engine := promql.NewEngine(promql.EngineOpts{
```
[the remark this answers is not in the document]
Comment (tomwilkie) [reply pkg/ruler/ruler_test.go]: Yes, moved to the "Update Cortex for changes in upstream Prometheus." cset.

File: pkg/configs/configs.go [hunk 3]
```diff
@@ -244,11 +246,23 @@ func (c RulesConfig) parseV1() (map[string][]rules.Rule, error) {
 			var rule rules.Rule
 
 			switch r := stmt.(type) {
-			case *promql.AlertStmt:
-				rule = rules.NewAlertingRule(r.Name, r.Expr, r.Duration, r.Labels, r.Annotations, true, util.Logger)
+			case *legacy_promql.AlertStmt:
+				// Re-parse the expression to get it into the right types.
```
[the remark this answers is not in the document]
Comment (tomwilkie) [reply pkg/configs/configs.go]: Not quite.  

`legacy_promql.ParseStmts` has parsed the whole rule for us. Ideally we'd just use `r.Expr` and pass that to `rules.NewAlertingRule`, but it is of the type `legacy_proql.Expr` and not `promql.Expr`.  So we convert it back to a string, and then parse it again with the upstream parser to get it into the right type.

I'll add that as the comment.

**Comment by tomwilkie** [comment]
> Do we have any tests that check V1 and V2 syntax do what we expect?

We do now...

**Pull request merged**